app/Services/Patient/VaccinationService.cs

```diff
@@ -0,0 +1,66 @@
+﻿using backend.Helpers;
+using backend.Database;
+using backend.Dto.Responses;
+using backend.Models.Visits;
+using backend.Exceptions;
+using backend.Models.Accounts;
+using backend.Models.Vaccines;
+
+namespace backend.Services.Patient
+{
+	public class VaccinationService
+	{
+		private readonly DataContext dataContext;
+        private readonly Mailer mailer;
+
+        public VaccinationService(DataContext dataContext, Mailer mailer)
+        {
+            this.dataContext = dataContext;
+            this.mailer = mailer;
+        }
+
+        public async Task<SuccessResponse> ReserveVaccinationSlot(PatientModel patient, int vaccinationSlotId, int vaccineId)
+        {
+            // Find matching vaccine
+            VaccineModel vaccine = this.dataContext.Vaccines
```

Comment:
czy moglibyśmy tutaj użyć semafora (https://docs.microsoft.com/pl-pl/dotnet/api/system.threading.semaphore?view=net-6.0)? wait one na początku i release po SaveChanges()? Wiem, że nasz serwer ma tylko 1 wątek, ale ktoś to może wyexploitować teoretycznie

Reply:
Bardzo dobry pomysł, już dodałem. 👍

Semaforem jest teraz zabezpieczona część funkcji od sprawdzania stanu slotu do zapisywania w bazie, tak, żeby ewentualne wyjątki przy _FirstOrThrow_ nie pozostawiły nam semafora w zablokowanym stanie.